### ISSUE TITLE
Parameterize the open flags when creating new devices

### ISSUE THREAD

**Issue by guygastineau** [georgefst/evdev#8]
https://github.com/georgefst/evdev/blob/a3959e51d3e4274ca96c945e81ef7cbb0453511e/evdev/src/Evdev/LowLevel.chs#L86

Hi, I want to use Haskell to replace a daemon that controls resources through libevdev in a work project.  The line of source code above has `defaultFlags` hard coded into the LowLevel module.  Obviously this is a problem when we need to set `O_NONBLOCK` (as is my need).

I am trying to figure out how to override this library with my local clone in Nix since there is no stack config for this project.  I am more than happy to supply a patch for this once I have it building my local copy with Nix.  Personally, I would prefer for `OpenFileFlags` to be an instance of `Default` from `Data.Default`.  If the `EvDev` module then re-exported `OpenFileFlags` as an instance of `Default` it would provide a simple yet flexible interface for `newDevice` I think.

Let me know if you would prefer that I approach this differently.  I don't want to dictate how your project should be structured ;)

BTW, aside from the `O_NONBLOCK` problem, this library has worked very well and easily for me.  Thank you very much for your effort :heart_eyes:

**Comment by georgefst** [comment]
This library tries to pull off a balancing act between providing most of the power of `libevdev`, and using a nice high-level Haskell-y interface. This certainly seems like somewhere where we currently fall short on the former, and I'm trying to remember why I designed it this way... I think I struggled to get any sensible behaviour with any other set of flags, but I don't remember the details. Have you verified that `O_NONBLOCK` does do what you expect?

We can easily add a `newDevice'` which allows the user to specify `OpenFileFlags`. Another option might be to allow the user to pass their own file descriptor, but we'd have to make some changes to enable this - currently, because the file descriptor is handled entirely by this library, we close it upon garbage collection (this approach is not always ideal anyway...).

I'm not that sold on the efficacy of a general `Default` type class. And especially seeing as it would incur a dependency on a library I'm not familiar with, and require an orphan instance, it's pretty much a no-go here.

And unfortunately, a Stack config in this project would just end up being unmaintained, since I don't use it. I think `cabal2nix` makes it fairly straightforward to add a dependency from a git repo, but I don't know how that integrates with Stack.

**Comment by guygastineau** [comment]
Hi, thank you for your response.  I think your perspectives are reasonable.

Yes, `O_NONBLOCK` does what I want in the C code, the Rust code, and now in the patched version of this library that I am running locally.

No stack is fine.  I actually really dislike Nix, but I finally installed Guix, and within 15 minutes I was able to get this building with it.  So, there are no worries there ;)

------

Honestly, I have been on a ~6-10 month hiatus from Haskell, where most of my work has been in Rust and Scheme dialects.  So, I am trying to remember to think more like a Haskeller for this again :)

I really like the idea of passing our own file descriptor (that way I can just easily use `Default` in my own projects without pushing that library on you :) ).  I imagine the signature for `newDevice` needs to stay the same to preserve the functionality of dependent code.

-----

I expect some users would prefer having access to `newDevice'` as you described it for the simplicity.  What do you think about adding `newDevice'` which takes an `OpenFileFlags` and then the more flexible (and basic) `newDeviceWithFd` which takes a file descriptor?

I can go ahead and start working on a PR for this if it sounds reasonable to you.  It sounds like a good bit of fun to get back into Haskell.

**Comment by guygastineau** [comment]
I realize that I didn't explain the behavior issues I was experiencing.  When using `evdev-streamly` to pretty print the events from a usb MSR (card reader that acts like a dumb keyboard), I get the first swipe, and then the stream dies.

**Comment by georgefst** [comment]
A PR adding both those variants would be much appreciated.

For `newDeviceFromFd`, you'd need to be wary of [the code which currently causes the file descriptors to be garbage collected](https://github.com/georgefst/evdev/commit/e286e5877215bef257062c87bdad63408e01faef). We'd need to store somewhere (on the C or Haskell side) a flag stating whether the `Fd` was provided by the user, and only close it in the finalizer if it was not. The haddocks in `newDevice` and `newDeviceFromFd` should make this behaviour clear to the library user.

**Comment by georgefst** [comment]
On reflection, I'm happy for `newDevice'` not to exist. Given that, as we've established, using different flags can cause odd behaviour, I think it's simpler to just have `newDeviceFromFd` as the version "for people who know what they're doing".

**Comment by guygastineau** [comment]
Sure that makes sense.

I will target adding `newDeviceFromFd` in a PR.  I have some other project deadlines approaching, so I don't expect to have a rough draft PR until early next week or the following the weekend.

Yeah, I am not entirely sure why I was loosing access to the file descriptor.  I am willing to spend some effort figuring it out though.

**Comment by guygastineau** [comment]
Hmm, sorry it has been a while since there has been any activity from me on this.

I might take another stab at it, but honestly I have wrapped my head around streamly a bit better now; I agree that I don't want non-blocking set up at the C level when using this Haskell library (especially with Streamly available).  If it could still benefit some users I don't mind trying to implement it though (I do like both C and Haskell ;) ), and I finally have my head wrapped around the ecosystem of build tools better now too.

**Comment by georgefst** [comment]
Ok cool. I'll leave this open, since there are probably other reasons why `newDeviceFromFd` could be useful to someone. But it's good to see that you're happily using the library as it is. I can imagine the non-blocking flag being more of a necessity in a lower-level language.

**Comment by guygastineau** [comment]
Hmm, I am having an issue that does relate to this somewhat again.

While trying to integrate this with streamly and gi-gtk-declarative there is a rather annoying issue.  I can link to an example program that highlights the issue too.

So, When streaming `S.repeatM . nextEvent` (like you have in your Evdev.Streamly) works like a charms, but `nextEvent` is like a black whole, and ending the program doesn't work.  I didn't realize this before, because `Ctrl-c` worked, but when trying to exit a GUI cleanly it hangs until another event is read from the device.  Also killing the program by doing `Ctrl-c` twice from a terminal that opened it works.  I tried using `asyncly` as in `asyncly . S.repeatM . nextEvent`, but then there are both repeated and dropped events sometimes.  This is similar to a situation I found myself in with a rust program I built that does similar work to what I am trying here.  In that situation I had to use a non-blocking file descriptor which had similar behavior of repeated/dropped events, so I had to do some sort of throttled reading to determine grouped bursts of events and ignore all the darn `EAGAIN` errno messages.

Basically, what I need this to do is stream lists of events split on newline key events, and I need the stream to terminate when asked instead of waiting until `nextEvent` returns again. I don't have my example project on GitHub, and it expects a card reader that acts like a keyboard (so it actually grabs the device).  Maybe it would be better if I just make a simpler minimal program for reproduction of the bug.

Anyway, I figured you might have an idea about how to solve this using your lib and Streamly.  If nothing else it is another gotcha for anyone using the library to note.  Thank you for your time as always :)

**Comment by guygastineau** [comment]
Hmm, it seems like a nextEventMay would suit my needs just fine (I could just filter the Nothings out).  I will mess around with making this function that returns a Nothing on `EAGAIN`.

**Comment by guygastineau** [comment]
So, I have fixed my problem, but I did so by creating both a `nextEventMay` and a `newDevice'` function.

I need to add some doc comments (and make an example?).

Anyway, you can see my ongoing work here if you want. https://github.com/guygastineau/evdev/tree/nextEventMay

I know before you said you didn't want `newDevice'` in the library, but this was much simpler to work out than including an unmanaged FD along with the default implementation.  And messing with the flags was easy.  For the record `nextEventMay` turns blocking off, but it has no effect for fd's that weren't created with `O_NONBLOCK`.  This also means that non blocking fds should work fine for the regular blocking reads.  That being said, there is no reason for someone to use custom flags when `defaultFileFlags` will do.

I feel a little weird about the fact that `nextEventMay` is not protecting against C errors with the `May`.  Maybe `Either Error (Maybe Event)` or something like that would be nice for a totally safe and recoverable error on a libevdev device.  Anyway, now I am just thinking publicly... :/

I will at least try to add doc comments to anything that is exported in my patch, and I will open a PR.  It might be easier to discuss it that way.

**Comment by georgefst** [comment]
I'd be interested in seeing that minimal reproducer. I've only used this lib for processes which were designed never to terminate, so there hasn't been a great deal of consideration of these sorts of issues.

One possibility that springs to mind is something like:
```hs
Streamly.mapMaybe id $ Streamly.takeWhile isJust $
    asyncly $
        (Just <$> readEvents dev)
            <>
        ((\() -> Nothing) <$> yieldM (takeMVar mvar))
```
where another thread would use `putMVar mvar ()` to end the stream.

> I feel a little weird about the fact that nextEventMay is not protecting against C errors with the May.

I wouldn't worry about that. This library happily [throws exceptions](https://github.com/guygastineau/evdev/blob/d341d8eb593849ccc230105b10ab7edf790ec799/evdev/src/Evdev.hs#L378) from `IO` actions. If you want an `Either Error a`, you can just wrap with [tryIOError](https://hackage.haskell.org/package/base-4.14.1.0/docs/System-IO-Error.html#v:tryIOError).

PS. I didn't realise that `Foreign.C.Error` exported those `Errno` constants, so thanks for that! I've made a note to simplify some existing code.

**Comment by guygastineau** [comment]
@georgefst 

Here is a run-able example of the behavior. https://github.com/guygastineau/evdev-streamly-blocking-bug/tree/nograb

If you check out the git history you will see some tags on that branch for `no hanging` (the desired behavior) and for `hanging` (the sort of buggy behavior).  I also have an explanation about how not-grabbing affects the behavior in the commit message for the `nograb-hang` tagged commit.

You can build any of the history from before the `nograb` branch, but beware unless you can close windows with a mouse too!

The project is the result of me dropping an Evdev example/idea into a simple hello-like example for `gi-gtk-declarative`.  The result is minimal and perhaps a little messy. Nevertheless, I think it is small and clear enough to showcase the blocking behavior of Evdev when the rest of the app wants to quit.

Finally, I tried to get it to build with `cabal new-build`, but I can't figure it out.  There is a branch I started dedicated to it (it is still device grabbing code), but I can't get the dependencies to resolve with new enough version of `haskell-gi` and `gi-glib`.  That is the result of some `gi-*` bug that is upstream of `gi-gtk` and `gi-gtk-declarative` is subject to those woes. Once I test out the newest stack resolver my PR that fixes this should get merged into `gi-gtk-declarative`.  Maybe that would make `cabal` happy?  IDK, `cabal` build that package from within its own git repository, so somehow I am just misusing `cabal` in this case.

Regardless, if you are willing to use `stack`, or if you can manage to build that exampel with `cabal`, then you can run the project.

**Comment by georgefst** [comment]
Thanks, I'll take a look at that when I can.

I have built `gi-gtk-declarative` pretty recently with no issues with Cabal 3 on Manjaro, but perhaps something has changed. Presumably if I hit issues, then [this](https://github.com/owickstrom/gi-gtk-declarative/pull/90) is the branch to try?

**Comment by guygastineau** [comment]
Yes, that branch should build if cabal can't build it from hackage.

**Comment by guygastineau** [comment]
Hmm, so today I have been messing around with things some more.  I just tried removing the `close` call in the finalizer, and it didn't seem to slow anything down when ending things.

For some reason I remember trying it last year, and it seemed like it slowed something down, but I don't know why.  The haskell runtime is pretty impressive, and it knows if a `Device` has the only reference to an Fd (I assume that is why you have the `Fd` in `Device` in the first place, ie. not only does it seem natural, but it tells the GC not to close the Fd)).

I know some docs say not to rely on the GC to close our file descriptors, but I don't think this is compelling.  In [this](https://stackoverflow.com/questions/10806753/when-is-it-better-to-explicitly-close-file-handles/10807279) stack exchange poster runs the following code and only ever has 70 file descriptors open at a time:

``` haskell
main = do
  mapM_ (\idx -> readIt) [1..1000]
  where readIt = do
          contents <- readFile "/etc/passwd"
          putChar $ head contents
```

From a note/comment in `LowLevel.hs` I surmise you read the warning in System.IO's haddock about not relying on the GC for this behavior.  Personally, I expect it not to be an issue.  You seem to have long running processes that don't end (on purpose I take it anyway), and I have processes that run all day long before being shutdown.  Maybe some use cases would stress this a bit more, but I think someone would really need to screw something up in order either to run out of fd's in a process or encounter some other issues from just letting the GC close the files.

That being said, I am also not against the idea of explicitly closing them (although it does seem strange to close the FD with `close` in C (does the haskell GC and runtime even know this has happened?). It still confuses me a bit, but it doesn't seem to cause any buggy behavior, so I am not really worried about it.

So, I think there are two sensible solutions going forward.

1. Let the GC collect all of the Fds for this library.
2. Eat all of the Fds, and warn users that supplying their own Fd means they can't rely on it.

In both of these cases Fd supplication up to the higher level module namely via a `newDeviceFromFd` function; `newDevice` (with its current behavior) would then be defined in terms of the more general `newDeviceFromFd`.

I think either of these outcomes is reasonable. I doubt anyone really needs their file descriptor after they have finished using the device through libevdev, so eating user provided Fds seems fine to me although there should be a warning.

Alternatively, I also think the GC is going to do a pretty good job of cleaning up the descriptors as evidenced from stack overflow and my own little experiments today.

Regardless of the approach we can agree on I think that they both fit a nice common and general interface.  In my branch I will go ahead and rework things to sit on top of `newDeviceFromFd`.  That means I can drop the constructor that takes open file options.  I will include the `nextEventMay` as the non-blocking version, since it requires more than just a change of the read flags.

I expect I will have it pretty cleaned up by later tonight or tomorrow (EST), and I will make a PR once I think it is ready.  I look forward to your review of the work ;)
